client/signup/steps/woocommerce-install/hooks/use-site-settings/index.tsx

```diff
@@ -0,0 +1,109 @@
+import { useCallback, useEffect, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import {
+	requestSiteSettings,
+	saveSiteSettings,
+	updateSiteSettings,
+} from 'calypso/state/site-settings/actions';
+import { getSiteSettings } from 'calypso/state/site-settings/selectors';
+
+// WooCommerce single options.
+export const WOOCOMMERCE_STORE_ADDRESS_1 = 'woocommerce_store_address';
+export const WOOCOMMERCE_STORE_ADDRESS_2 = 'woocommerce_store_address_2';
+export const WOOCOMMERCE_STORE_CITY = 'woocommerce_store_city';
+export const WOOCOMMERCE_DEFAULT_COUNTRY = 'woocommerce_default_country';
+export const WOOCOMMERCE_STORE_POSTCODE = 'woocommerce_store_postcode';
+
+// Map Woop to WooCoommerce single options.
+const siteOptionsMap = {
+	address_1: WOOCOMMERCE_STORE_ADDRESS_1,
+	address_2: WOOCOMMERCE_STORE_ADDRESS_2,
+	city: WOOCOMMERCE_STORE_CITY,
+	postcode: WOOCOMMERCE_STORE_POSTCODE,
+	country: WOOCOMMERCE_DEFAULT_COUNTRY,
+};
+
+/**
+ * Returns the option name for the given WooCommerce option.
+ *
+ * @param {string} option option name
+ * @returns {string} option
+ */
+function getSiteOptionName( option: string ) {
+	if ( ! ( option in siteOptionsMap ) ) {
+		return option;
+	}
+
+	return siteOptionsMap[ option ];
+}
+
+/**
+ * Simple react custom hook to deal with site settings.
+ *
+ * @param {number} siteId - site id
+ * @returns {Array} - site option handlers
+ */
+export function useSiteSettings( siteId: number ) {
+	const dispatch = useDispatch();
+
+	const settings = useSelector( ( state ) => getSiteSettings( state, siteId ) );
+
+	/*
+	 * Private settings store.
+	 * It collects the options that will be updated/saved
+	 */
+	const [ editedSettings, setEditedSettings ] = useState( {} );
+
+	// Dispatch action to request the site settings.
+	useEffect( () => {
+		if ( ! siteId ) {
+			return;
+		}
+
+		dispatch( requestSiteSettings( siteId ) );
+	}, [ dispatch, siteId ] );
+
+	// Simple getter helper.
+	function get( option: string ) {
+		if ( ! settings || Object.keys( settings ).length === 0 ) {
+			return '';
+		}
+
+		const value = settings?.[ getSiteOptionName( option ) ] || '';
+		return value;
+	}
+
+	/*
+	 * Helper to 'update' site settings data.
+	 * Chnages are applied to the Redux store.
+	 */
+	const update = useCallback(
+		( option: string, value: string ) => {
+			const siteOption = getSiteOptionName( option );
+			setEditedSettings( ( state ) => ( { ...state, [ siteOption ]: value } ) );
```

Comment:
Why are this settings being saved in the redux store and also a private store?

Reply:
Nice question, Rolando.
We need to track what are the options that we are going to update.
If the user fills the address 1, address, 2, city... etc, we need to collect them in order to pass what are the final values that we want to save.
The [update](https://github.com/Automattic/wp-calypso/blob/5166a2a25e5ce8b94e882dd0a32847f687cb86b5/client/state/site-settings/actions.js#L39) method only updates the redux store, while the [save](https://github.com/Automattic/wp-calypso/blob/5166a2a25e5ce8b94e882dd0a32847f687cb86b5/client/state/site-settings/actions.js#L39) method performs the POST endpoint request.
But the [save method](https://github.com/Automattic/wp-calypso/blob/5166a2a25e5ce8b94e882dd0a32847f687cb86b5/client/state/site-settings/actions.js#L86) ALSO expects a settings object as the second argument, and that is the private store (useState) what we pass to it.
In an Ideal World, we shouldn't take care about what are the data that we want to save since it should be tracked by redux when the `update` method is called. We simply should call `save()` without passing any settings data.

Reply:
Makes sense, I wonder if instead we can use the private store to just track which attributes are being updated (address 1, address 2) as a string and then pick them up from the redux state. Maybe it can be a more descriptive way of declaring the private store?
```
// on declaration
const [ editedSettings, setEditedSettings ] = useState([]);

// on update
setEditedSettings([ ...editedSettings,  'address1']);

// on save
const settingsToSave = editedSettings.map(key => settings[key]);
```

Reply:
that makes totally sense. I'd suggest defining a function to deal with updating the state in the hook, instead of calling it value. Something like:

```es6
setEditedSettings( ( settings ) => ( [ ...settings,  'address1'] ) );
```

Also, check we don't duplicate items into the Array.

Reply:
Good catch! there! we totally need to check for uniqueness too!

Reply:
Addressed here: https://github.com/Automattic/wp-calypso/pull/59505/commits/0621f7369c2e38e374f3e3d0cf499e59053601f3